core/src/main/java/org/infinispan/stream/impl/ClusterStreamManagerImpl.java

```diff
@@ -0,0 +1,438 @@
+                  Set<K> keysExcluded = determineExcludedKeys(keysToExclude, targetSegments);
+                  log.tracef("Submitting task to %s for %s excluding keys %s", dest, uuid, keysExcluded);
+                  Response response = rpc.invokeRemotely(Collections.singleton(dest), factory.buildStreamRequestCommand(
+                          uuid, parallelStream, StreamRequestCommand.Type.TERMINAL_KEY_REHASH, targetSegments,
+                          keysToInclude, keysExcluded, includeLoader, operation),
+                          rpc.getDefaultRpcOptions(true)).values().iterator().next();
+                  if (!response.isSuccessful()) {
+                     log.tracef("Unsuccessful response for %s from %s - making segments %s suspect", uuid,
+                             dest, targetSegments);
+                     receiveResponse(uuid, dest, true, targetSegments, null);
+                  }
+               } catch (Exception e) {
+                  Throwable cause = e;
+                  boolean wasSuspect = false;
+                  // Unwrap exception
+                  do {
+                     if (cause instanceof SuspectException) {
+                        log.tracef("Exception from %s contained a SuspectException, making all segments %s suspect",
+                                dest, targetSegments);
+                        receiveResponse(uuid, dest, true, targetSegments, null);
+                        wasSuspect = true;
+                        break;
+                     }
+                  } while ((cause = cause.getCause()) != null);
+
+                  if (!wasSuspect) {
+                     log.tracef(e, "Encounted exception for %s from %s", uuid, dest);
+                     throw e;
+                  }
+               }
+            }
+         }
+      }
+      return uuid;
+   }
+
+   private void submitAsyncTasks(UUID uuid, Map<Address, Set<Integer>> targets, Map<Integer, Set<K>> keysToExclude,
+                                 boolean parallelStream, Set<K> keysToInclude, boolean includeLoader,
+                                 StreamRequestCommand.Type type, Object operation) {
+      for (Map.Entry<Address, Set<Integer>> targetInfo : targets.entrySet()) {
+         Set<Integer> segments = targetInfo.getValue();
+         Set<K> keysExcluded = determineExcludedKeys(keysToExclude, segments);
+         Address dest = targetInfo.getKey();
+         log.tracef("Submitting async task to %s for %s excluding keys %s", dest, uuid, keysExcluded);
+         CompletableFuture<Map<Address, Response>> completableFuture = rpc.invokeRemotelyAsync(
+                 Collections.singleton(dest), factory.buildStreamRequestCommand(uuid, parallelStream, type, segments,
+                         keysToInclude, keysExcluded, includeLoader, operation),
+                 rpc.getDefaultRpcOptions(true));
+         completableFuture.whenComplete((v, e) -> {
```

Comment:
If `submitAsyncTasks` worked on a single destination and returned this future, I think you could use it for the non-parallelDistribution case as well.